src/Setup_geometry.jl

```diff
@@ -447,6 +447,67 @@ function AddCylinder!(Phase, Temp, Grid::AbstractGeneralGrid;   # required input
     return nothing
 end
 
+
+
+function AddPolygon!(Phase, Temp, Grid::AbstractGeneralGrid;                 # required input
+    xlim=Tuple{}, ylim=Tuple{2}, zlim=Tuple{},     # limits of the box
+    Origin=nothing, StrikeAngle=0, DipAngle=0,      # origin & dip/strike
+    phase = ConstantPhase(1),                       # Sets the phase number(s) in the box
+    T=nothing )                                     # Sets the thermal structure (various functions are available)
+
+# Retrieve 3D data arrays for the grid
+X,Y,Z = coordinate_grids(Grid)
+
+
+indx = zeros(length(X))
+
+# find points of the total script within the polygone, only in 2D due to the symetric structures and index of y
+for i = 1:length(X)
+
+    # working but not the fastest
+    if Y[i] >= ylim[1] && Y[i]<=ylim[2] 
+        indx[i] = inPoly(xlim,zlim, X[i], Z[i])
+    end
+end
+
+# get all indices which are in the polygone separated
+ind = findall(x->x>0,indx)
+
+
+# Compute thermal structure accordingly. See routines below for different options
+if T != nothing
+Temp[ind] = Compute_ThermalStructure(Temp[ind], X[ind], Y[ind], Z[ind], T)
+end
+
+# Set the phase. Different routines are available for that - see below.
+Phase[ind] = Compute_Phase(Phase[ind], Temp[ind], X[ind], Y[ind], Z[ind], phase)
+
+return nothing
+end
+
+
+
+function inPoly(PolyX, PolyY, x, y)
+    inside = false
+    iSteps = collect(eachindex(PolyX))
```

Comment:
Do you need to `collect` here? it will allocate an array

Reply:
I often default to using collect because I am never sure what I can and can't do with the specific julia types. if `eachindex(PolyX)` doesn't iterate, then we can get rid of `collect`

How can I improve the next line?

Reply:
If i understand correctly, `jsteps` is a circular array. You could do something like (check whether it is correct):

```julia
n = length(PolyX)
for i in eachindex(PolyX)
        j   = i-1
        j += n * (j<1)
        xi = PolyX[i]
        yi = PolyY[i]
        xj = PolyX[j]
        yj = PolyY[j]

        if ((yi > y) != (yj > y)) && (x < (xj - xi) * (y - yi) / (yj - yi + eps()) + xi)

            inside = !inside
        end
    end
```

Reply:
This works and gets rid of the last allocations. Thanks!

I already put the function in main yesterday, will make another PR with the update

Reply:
Just remember that `eachindex` and `axes` return (non-allocating) iterable objects, so you should  never `collect` them.